akka-http-tests/src/test/scala/akka/http/scaladsl/server/DontLeakActorsOnFailingConnectionSpecs.scala

```diff
@@ -42,11 +42,12 @@ abstract class DontLeakActorsOnFailingConnectionSpecs(poolImplementation: String
   "Http.superPool" should {
 
     "not leak connection Actors when hitting non-existing endpoint" in {
+      val address = SocketUtil.temporaryServerAddress()
```

Comment:
Was that an actual problem on your machine? Because I changed it from that because `temporaryServerAddress` isn't guaranteeing that there actually is no server when we try to connect.

Reply:
Yep, fails consistently on MacOS it seems.

Reply:
Seems like people have the same problem on windows, but I don't have a window so I can try.

Reply:
Searching a bit it seems as only 127.0.0.1 is mapped to loopback by default in MacOS and any additional IP in the 127.0.0.0/8 to be looped back has to be explicitly aliased :/

Reply:
Yes, but what's the error? The point of the test is to connect to a random unconnectable address?

Reply:
Ah, failed to realize that, my bad, let's see.

Reply:
With the original code the assertion just fails `countDown.await(10, TimeUnit.SECONDS) should be(true)`, nothing logged.

Reply:
It should also show those printlns:

```
2: FAIL: akka.stream.StreamTcpException: Tcp command [Connect(127.0.0.46:34763,None,List(),Some(10 seconds),true)] failed because of Connection refused
```

Does it do that? Or I guess what could be happening is that the connection attempt just runs into a timeout? Could you try to telnet to that address and see what happens?

Reply:
Here's the debug log output in case you can spot anything obvious:

```
akka-http-root > akka-http-tests/testOnly akka.http.scaladsl.server.NewPoolDontLeakActorsOnFailingConnectionSpecs
[DEBUG] [01/15/2018 16:25:25.764] [pool-66-thread-10] [EventStream] StandardOutLogger started
[DEBUG] [01/15/2018 16:25:25.906] [pool-66-thread-10] [EventStream(akka://DontLeakActorsOnFailingConnectionSpecs-new)] logger log1-TestEventListener started
[DEBUG] [01/15/2018 16:25:25.907] [pool-66-thread-10] [EventStream(akka://DontLeakActorsOnFailingConnectionSpecs-new)] Default Loggers started
[DEBUG] [01/15/2018 16:25:25.906] [pool-66-thread-10] [EventStream(akka://DontLeakActorsOnFailingConnectionSpecs-new)] logger log1-TestEventListener started
[DEBUG] [01/15/2018 16:25:25.907] [pool-66-thread-10] [EventStream(akka://DontLeakActorsOnFailingConnectionSpecs-new)] Default Loggers started
[DEBUG] [01/15/2018 16:25:26.066] [pool-66-thread-10-ScalaTest-running-NewPoolDontLeakActorsOnFailingConnectionSpecs] [AkkaSSLConfig(akka://DontLeakActorsOnFailingConnectionSpecs-new)] Initializing AkkaSSLConfig extension...
[DEBUG] [01/15/2018 16:25:26.068] [pool-66-thread-10-ScalaTest-running-NewPoolDontLeakActorsOnFailingConnectionSpecs] [AkkaSSLConfig(akka://DontLeakActorsOnFailingConnectionSpecs-new)] buildHostnameVerifier: created hostname verifier: com.typesafe.sslconfig.ssl.DefaultHostnameVerifier@cfda3f4
[DEBUG] [01/15/2018 16:25:26.360] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [akka.stream.Log(akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-0)] [whatever] Element: (HttpRequest(HttpMethod(GET),http://127.0.0.46:34763/test/1,List(),HttpEntity.Strict(none/none,ByteString()),HttpProtocol(HTTP/1.1)),1)
[DEBUG] [01/15/2018 16:25:26.365] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [akka.stream.Log(akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-0)] [whatever] Element: (HttpRequest(HttpMethod(GET),http://127.0.0.46:34763/test/2,List(),HttpEntity.Strict(none/none,ByteString()),HttpProtocol(HTTP/1.1)),2)
[DEBUG] [01/15/2018 16:25:26.365] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [akka.stream.Log(akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-0)] [whatever] Element: (HttpRequest(HttpMethod(GET),http://127.0.0.46:34763/test/3,List(),HttpEntity.Strict(none/none,ByteString()),HttpProtocol(HTTP/1.1)),3)
[DEBUG] [01/15/2018 16:25:26.365] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [akka.stream.Log(akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-0)] [whatever] Element: (HttpRequest(HttpMethod(GET),http://127.0.0.46:34763/test/4,List(),HttpEntity.Strict(none/none,ByteString()),HttpProtocol(HTTP/1.1)),4)
[DEBUG] [01/15/2018 16:25:26.372] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] (Re-)starting host connection pool to 127.0.0.46:34763
[DEBUG] [01/15/2018 16:25:26.490] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] InputBuffer (max-open-requests = 32) now filled with 1 request after enqueuing GET /test/1 Empty
[DEBUG] [01/15/2018 16:25:26.490] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] InputBuffer (max-open-requests = 32) now filled with 2 request after enqueuing GET /test/2 Empty
[DEBUG] [01/15/2018 16:25:26.490] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] InputBuffer (max-open-requests = 32) now filled with 3 request after enqueuing GET /test/3 Empty
[DEBUG] [01/15/2018 16:25:26.491] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] InputBuffer (max-open-requests = 32) now filled with 4 request after enqueuing GET /test/4 Empty
[DEBUG] [01/15/2018 16:25:26.495] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [0 (Unconnected)] Dispatching request
[DEBUG] [01/15/2018 16:25:26.500] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [0 (Unconnected)] Establishing connection
[DEBUG] [01/15/2018 16:25:26.519] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [0 (Connecting)] State change [Unconnected] -> [Connecting]
[DEBUG] [01/15/2018 16:25:26.535] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [1 (Unconnected)] Dispatching request
[DEBUG] [01/15/2018 16:25:26.535] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [1 (Unconnected)] Establishing connection
[DEBUG] [01/15/2018 16:25:26.535] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [1 (Connecting)] State change [Unconnected] -> [Connecting]
[DEBUG] [01/15/2018 16:25:26.536] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [2 (Unconnected)] Dispatching request
[DEBUG] [01/15/2018 16:25:26.536] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [2 (Unconnected)] Establishing connection
[DEBUG] [01/15/2018 16:25:26.537] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-6] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [2 (Connecting)] State change [Unconnected] -> [Connecting]
[DEBUG] [01/15/2018 16:25:26.537] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [3 (Unconnected)] Dispatching request
[DEBUG] [01/15/2018 16:25:26.537] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [3 (Unconnected)] Establishing connection
[DEBUG] [01/15/2018 16:25:26.538] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [3 (Connecting)] State change [Unconnected] -> [Connecting]
[DEBUG] [01/15/2018 16:25:26.538] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-2] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/0] Resolving 127.0.0.46 before connecting
[DEBUG] [01/15/2018 16:25:26.538] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-5] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/1] Resolving 127.0.0.46 before connecting
[DEBUG] [01/15/2018 16:25:26.541] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-8] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/2] Resolving 127.0.0.46 before connecting
[DEBUG] [01/15/2018 16:25:26.541] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-9] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/3] Resolving 127.0.0.46 before connecting
[DEBUG] [01/15/2018 16:25:26.552] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-3] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-DNS] Resolution request for 127.0.0.46 from Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/0#-1107223284]
[DEBUG] [01/15/2018 16:25:26.556] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-3] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-DNS] Resolution request for 127.0.0.46 from Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/1#924546716]
[DEBUG] [01/15/2018 16:25:26.557] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-3] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-DNS] Resolution request for 127.0.0.46 from Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/3#617602379]
[DEBUG] [01/15/2018 16:25:26.557] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-3] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-DNS] Resolution request for 127.0.0.46 from Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/2#280529127]
[DEBUG] [01/15/2018 16:25:26.561] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-3] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/1] Attempting connection to [/127.0.0.46:34763]
[DEBUG] [01/15/2018 16:25:26.561] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-5] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/0] Attempting connection to [/127.0.0.46:34763]
[DEBUG] [01/15/2018 16:25:26.561] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-8] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/2] Attempting connection to [/127.0.0.46:34763]
[DEBUG] [01/15/2018 16:25:26.561] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-2] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/3] Attempting connection to [/127.0.0.46:34763]
[DEBUG] [01/15/2018 16:25:36.348] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] Pool upstream failed with akka.stream.AbruptTerminationException: Processor actor [Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-1/flow-1-0-PoolFlow#-1169168432]] terminated abruptly
[DEBUG] [01/15/2018 16:25:36.348] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] Pool stopped
[WARN] [01/15/2018 16:25:36.351] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [0 (Connecting)] Ongoing request [GET /test/1 Empty] was dropped because pool is shutting down
[WARN] [01/15/2018 16:25:36.351] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [1 (Connecting)] Ongoing request [GET /test/2 Empty] was dropped because pool is shutting down
[WARN] [01/15/2018 16:25:36.351] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [2 (Connecting)] Ongoing request [GET /test/3 Empty] was dropped because pool is shutting down
[WARN] [01/15/2018 16:25:36.351] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-4] [DontLeakActorsOnFailingConnectionSpecs-new/Pool(shared->http://127.0.0.46:34763)] [3 (Connecting)] Ongoing request [GET /test/4 Empty] was dropped because pool is shutting down
[DEBUG] [01/15/2018 16:25:36.351] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-2] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/1] Monitored actor [Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-1/$$b#197619218]] terminated
[DEBUG] [01/15/2018 16:25:36.351] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-2] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/2] Monitored actor [Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-1/$$c#-1423111567]] terminated
[DEBUG] [01/15/2018 16:25:36.353] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-8] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/0] Monitored actor [Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-1/$$a#368051911]] terminated
[DEBUG] [01/15/2018 16:25:36.353] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-8] [akka://DontLeakActorsOnFailingConnectionSpecs-new/system/IO-TCP/selectors/$a/3] Monitored actor [Actor[akka://DontLeakActorsOnFailingConnectionSpecs-new/user/StreamSupervisor-1/$$d#1455268611]] terminated
[DEBUG] [01/15/2018 16:25:36.357] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-8] [EventStream] shutting down: StandardOutLogger started
[DEBUG] [01/15/2018 16:25:36.357] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-8] [EventStream] shutting down: StandardOutLogger started
[DEBUG] [01/15/2018 16:25:36.358] [DontLeakActorsOnFailingConnectionSpecs-new-akka.actor.default-dispatcher-8] [EventStream] all default loggers stopped
[info] NewPoolDontLeakActorsOnFailingConnectionSpecs:
[info] Http.superPool
[info] - should not leak connection Actors when hitting non-existing endpoint *** FAILED ***
[info]   false was not true (DontLeakActorsOnFailingConnectionSpecs.scala:62)
```

Reply:
I can see a similar thing if I change to `128.0...`.

Reply:
Manually connecting to 127.0.0.46 with telnet just hangs forever, so there is no such ip but there is a route to it O.o

No logged things from the `foreach` there are no elements reaching it.

Reply:
It actually might be a bug...

Reply:
Maybe related to that one: https://github.com/akka/akka-http/issues/1736?

It seems a connection is not being reused after it a got a single connection error.

Reply:
Note that I see it both for the legacy and the new connection pool.

Reply:
> Note that I see it both for the legacy and the new connection pool.

Same for me.

What I see in wireshark is, that if I put in a "weird address", it gets routed according to the default route (which usually is the internet route). For the first few connection attempts I get back a fast response (from some router, probably the first ISP one). Eventually, after a few more requests, it seems that this router seems to lose or drop that SYN packet and also a few more of the SYN resends. However, automatic connection re-attempts by the OS are using an exponential backoff, i.e. it only gets back the next response 4 SYN packets (or 8 seconds) later. That will then fail the test for me because roughly every 10th to 15th connection attempt will take >8 seconds.

Reply:
From your logs it seems you never even get the "No route to host" message back. Hard to say why that's the case but maybe let's keep your change for now (@raboof also argued for keeping the old code (= your suggestion) when I changed it a few weeks back).